crates/pallet-subspace/src/lib.rs

```diff
@@ -810,6 +1010,321 @@ impl<T: Config> Pallet<T> {
+    InvalidSolution(VerificationError<Header>),
+    Equivocated,
+}
+
+impl<Header> From<CheckVoteError<Header>> for TransactionValidityError
+where
+    Header: HeaderT,
+{
+    fn from(error: CheckVoteError<Header>) -> Self {
+        TransactionValidityError::Invalid(match error {
+            CheckVoteError::BlockListed => InvalidTransaction::BadSigner,
+            CheckVoteError::UnexpectedBeforeHeightTwo => InvalidTransaction::Call,
+            CheckVoteError::HeightInTheFuture => InvalidTransaction::Future,
+            CheckVoteError::HeightInThePast => InvalidTransaction::Stale,
+            CheckVoteError::IncorrectParentHash => InvalidTransaction::Call,
+            CheckVoteError::SlotInTheFuture => InvalidTransaction::Future,
+            CheckVoteError::SlotInThePast => InvalidTransaction::Stale,
+            CheckVoteError::BadRewardSignature(_) => InvalidTransaction::BadProof,
+            CheckVoteError::UnknownRecordsRoot => InvalidTransaction::Call,
+            CheckVoteError::InvalidSolution(_) => InvalidTransaction::Call,
+            CheckVoteError::Equivocated => InvalidTransaction::BadSigner,
+        })
+    }
+}
+
+fn check_vote<T: Config>(
+    signed_vote: &SignedVote<T::BlockNumber, T::Hash, T::AccountId>,
+    pre_dispatch: bool,
+) -> Result<(), CheckVoteError<T::Header>> {
+    let Vote::V0 {
+        height,
+        parent_hash,
+        slot,
+        solution,
+    } = &signed_vote.vote;
+    let height = *height;
+    let slot = *slot;
+
+    if BlockList::<T>::contains_key(&solution.public_key) {
+        return Err(CheckVoteError::BlockListed);
+    }
+
+    let current_block_number = frame_system::Pallet::<T>::current_block_number();
+
+    if current_block_number <= One::one() || height <= One::one() {
+        debug!(
+            target: "runtime::subspace",
+            "Votes are not expected at height below 2"
+        );
+
+        return Err(CheckVoteError::UnexpectedBeforeHeightTwo);
+    }
+
+    // Height must be either the same as in current block or smaller by one.
+    //
+    // Subtraction will not panic due to check above.
+    if !(height == current_block_number || height == current_block_number - One::one()) {
+        debug!(
+            target: "runtime::subspace",
+            "Vote verification error: bad height {height:?}, current block number is \
+            {current_block_number:?}"
+        );
+        return Err(if height > current_block_number {
+            CheckVoteError::HeightInTheFuture
+        } else {
+            CheckVoteError::HeightInThePast
+        });
+    }
+
+    // Should have parent hash from -1 (parent hash of current block) or -2 (block before that)
+    //
+    // Subtraction will not panic due to check above.
+    if *parent_hash != frame_system::Pallet::<T>::block_hash(height - One::one()) {
+        debug!(
+            target: "runtime::subspace",
+            "Vote verification error: parent hash {}",
+            hex::encode(parent_hash)
```

Comment:
Why have to encode `parent_hash` in hex again? It's already in hex with Display/Debug format.